### ISSUE TITLE
Question about Quantization-int8

### ISSUE THREAD

**Issue by ybai62868** [pytorch/glow#4420]
Hi,

I want to use the quantization in Glow. When I use this command:
../bin/image-classifier ../tests/images/imagenet/zebra_340.png -image-mode=0to1 -m=../resnet50/model.onnx -model-input-name=gpu_0/data_0,[int8q,[0.123,-13],[1,10]] -dump-profile="profile.yaml"

I got the error as following:
0415 22:32:16.624437 189784064 ONNXModelLoader.cpp:3392] Error while trying to ConstantFold Conv_gpu_0/conv1_1:
Error message: No node under name gpu_0/data_0
Error return stack:
../lib/Importer/ProtobufLoader.cpp:175
../lib/Importer/ONNXModelLoader.cpp:2045
F0415 22:32:16.638357 189784064 Error.cpp:119] exitOnError(Error) got an unexpected ErrorValue:
Error message: No node under name gpu_0/data_0
Error return stack:
../lib/Importer/ProtobufLoader.cpp:175
../lib/Importer/ONNXModelLoader.cpp:3166
../lib/Importer/ONNXModelLoader.cpp:3399
../lib/Importer/ONNXModelLoader.cpp:3525
../lib/Importer/ONNXModelLoader.cpp:3539
*** Check failure stack trace: ***
0  image-classifier         0x000000010f8511ac llvm::sys::PrintStackTrace(llvm::raw_ostream&) + 37
1  image-classifier         0x000000010f8515de SignalHandler(int) + 192
2  libsystem_platform.dylib 0x00007fff76423b5d _sigtramp + 29
3  libsystem_platform.dylib 000000000000000000 _sigtramp + 2310915264
4  libsystem_c.dylib        0x00007fff762dd6a6 abort + 127
5  libglog.0.3.5.dylib      0x0000000112e9eca9 google::InstallFailureFunction(void (*)()) + 0
6  libglog.0.3.5.dylib      0x0000000112e9eca0 google::logging_fail() + 0
7  libglog.0.3.5.dylib      0x0000000112e9defd google::LogMessage::SendToLog() + 717
8  libglog.0.3.5.dylib      0x0000000112e9e677 google::LogMessage::Flush() + 175
9  libglog.0.3.5.dylib      0x0000000112ea1bf7 google::LogMessageFatal::~LogMessageFatal() + 15
10 libglog.0.3.5.dylib      0x0000000112e9ef89 google::LogMessageFatal::~LogMessageFatal() + 9
11 image-classifier         0x0000000111124633 glow::detail::exitOnError(char const*, unsigned long, glow::detail::GlowError) + 387
12 image-classifier         0x000000010f734ccc glow::ONNXModelLoader::ONNXModelLoader(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, llvm::ArrayRef<char const*>, llvm::ArrayRef<glow::Type const*>, glow::Function&, glow::detail::GlowError*, bool, bool, glow::Backend const*) + 364
13 image-classifier         0x000000010f3cf9e4 buildAndCompileAndGetInAndOutPair(glow::Loader&, glow::PlaceholderBindings&, glow::Type const*) + 900
14 image-classifier         0x000000010f3cc004 void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, glow::Executor::executeNetwork(int, char**)::$_1> >(void*) + 2564
15 libsystem_pthread.dylib  0x00007fff7642c2eb _pthread_body + 126
16 libsystem_pthread.dylib  0x00007fff7642f249 _pthread_start + 66
17 libsystem_pthread.dylib  0x00007fff7642b40d thread_start + 13
profiling.sh: line 1: 74801 Abort trap: 6           ../bin/image-classifier ../tests/images/imagenet/zebra_340.png -image-mode=0to1 -m=../resnet50/model.onnx -model-input-name=gpu_0/data_0,[int8q,[0.123,-13],[1,10]] -dump-profile="profile.yaml"
-----
But when I use the command:
../bin/image-classifier ../tests/images/imagenet/zebra_340.png -image-mode=0to1 -m=../resnet50/model.onnx -model-input-name=gpu_0/data_0 -dump-profile="profile.yaml"

It can run successfully and generate the profile.yaml file. If I want to use the "type,[scale,offset],shape]" for the quantization, what should I do?

Thanks

**Comment by mciprian13** [comment]
You can find more details here [AOT.md](https://github.com/pytorch/glow/blob/master/docs/AOT.md).
For generating a quantized bundle you need to do two steps:
1. Profile the model using the **image-classifier** to gather profiling data required for quantization. Profiling means running the model in floating-point version to gather data about the floating-point dynamic range of the tensors.
2. Compile the model into a bundle using the **model-compiler** tool.

Therefore you will use the profiling command like this:
`../bin/image-classifier ../tests/images/imagenet/zebra_340.png -image-mode=0to1 -model=../resnet50/model.onnx -model-input-name=gpu_0/data_0 -dump-profile="profile.yaml"`
Be aware that the parameter **-model-input-name** only provides the name of the model input placeholder and not its type.

Compile the model into a bundle (library/object) like this:
`../bin/model-compiler -model=../resnet50/model.onnx -load-profile="profile.yaml" -backend=CPU -emit-bundle=bundle`

Note that when compiling ONNX models with **model-compiler** you don't need to provide information about the input placeholder since it is derived automatically. Only for Caffe2 models you need to provide the parameter **-model-input=<name>,<type>,<shape>** and this specifies the name/types of the original model and not the quantized version.

The model is quantized implicitly through the use of the "-load-profile" parameter and you don't need to use anything else. Please read entirely the documentation I referenced since you will find there additional information and parameters.

**Comment by ybai62868** [comment]
Hi @mciprian13 ,
Thanks for your help! You tell me lots of things about using Glow to quantize the resnet50 model.
I read the documentation you mentioned before. I can successfully get the file which contains model.h, model.o,model.weights.bin and model.weights.txt. 
But I don't know how to combine all of them to my project to get the result of resnet50 model after quantization. I just find the "How to use the bundle" in [AOT.md](https://github.com/pytorch/glow/blob/master/docs/AOT.md). Do you know how to find some code example about using the bundle?

Once again, thank you!

**Comment by mciprian13** [comment]
The bundle is a library and you need to create an application (main.cpp) to integrate your bundle and run an inference.
You can find an application example for Resnet50 [here](https://github.com/pytorch/glow/tree/master/examples/bundles/lenet_mnist) and the associated CMake file to build. The Resnet50 bundle for this example was obtained from a Caffe2 model.
In the document I referenced are also plenty of details in how to integrate the bundle.
As stated in the documentation if you build Glow with the CMake option `-DGLOW_WITH_BUNDLES=ON` it will build the Resnet50 example automatically (both floating-point and quantized version).

**Comment by ybai62868** [comment]
@mciprian13 
Hi, I find the resnet50 and letnet example in examples/bundles, thanks a lot!

**Comment by ybai62868** [comment]
@mciprian13 
Hi, after reading all of the documentation you mentioned before it generates the resnet.weights.txt successfully. I use vim to open this file and find that have the following format:
0X25, 0X5F, 0X00, 0X00, 0X80
It seems like the weights after the int8 quantization. But I want to find the correspondence with the original resnet50 (before quantization). I mean which number can represent the weight of the first conv-layer. Because I want to compare the glow quantization result with another strategy.
Thanks a lot.

**Comment by mciprian13** [comment]
You could wait for the PR #4392 to be over (reviewed and merged) so that when you run the inference in instrumentation mode the bundle will dump all the tensors in separate files (including weights and intermediate tensors from the graph).

**Comment by ybai62868** [comment]
@mciprian13 Hi, I find that your PR has been merged into the master branch. I use the latest version of glow to build the project for the int8 quantization resnet50 model. I only find the update of the profile.yaml. But how can I get the correspondence with the original resnet50 (before the quantization and after the quantization when I mentioned before)?
Thanks a lot!

**Comment by ybai62868** [comment]
@mciprian13 Hi， I have successfully used the new function of your PR when I do the int8 quantization resnet50. 
I use the following command:

 ../bin/model-compiler -load-profile=profile.yaml -assert-all-nodes-quantized -keep-original-precision-for-nodes=SoftMax -backend=CPU -model=../resnet50/model.onnx -emit-bundle=build -model-input=gpu_0    /data_0,float,[1,3,224,224] -instrument-debug=true -instrument-debug-format=txt

And I find the debug.info in the debug folder, I open the debug.info and all of the information about it is as follow:
1   Format: txt
  1
  2 Type: quantize
  3 Name: Conv_gpu_0_conv1_1__1_quantize
  4 [0] Dest:      data0000.txt    i8[S:0.0039 O:-128][0.000,1.000]<1 x 3 x 224 x 224>
  5 [1] Src:       data0001.txt    float<1 x 3 x 224 x 224>
  6
  7 Type: transpose
  8 Name: Conv_gpu_0_conv1_1__7
  9 [0] Dest:      data0002.txt    i8[S:0.0039 O:-128][0.000,1.000]<1 x 224 x 224 x 3>
 10 [1] Src:       data0003.txt    i8[S:0.0039 O:-128][0.000,1.000]<1 x 3 x 224 x 224>
 11
 12 Type: convolution
 13 Name: Conv_gpu_0_conv1_1__8
 14 [0] Dest:      data0004.txt    i8[S:0.0318 O:-128][0.000,8.101]<1 x 112 x 112 x 64>
 15 [1] Src:       data0005.txt    i8[S:0.0039 O:-128][0.000,1.000]<1 x 224 x 224 x 3>
 16 [2] Filter:    data0006.txt    i8[S:0.0092 O:-4][-1.136,1.200]<64 x 7 x 7 x 3>
 17 [3] Bias:      data0007.txt    i32[S:0.0000 O:0][-77147.078,77147.078]<64>
 18
 19 Type: cpumaxsplat
 20 Name: Relu_gpu_0_res_conv1_bn_2__1_max__2
 21 [0] Dest:      data0008.txt    i8[S:0.0318 O:-128][0.000,8.101]<1 x 112 x 112 x 64>
 22 [1] Src:       data0009.txt    i8[S:0.0318 O:-128][0.000,8.101]<1 x 112 x 112 x 64>
 23
 24 Type: maxpool
 25 Name: MaxPool_gpu_0_pool1_1__2
 26 [0] Dest:      data0010.txt    i8[S:0.0318 O:-128][0.000,8.101]<1 x 56 x 56 x 64>
 27 [1] Src:       data0011.txt    i8[S:0.0318 O:-128][0.000,8.101]<1 x 112 x 112 x 64>
 28
 29 Type: convolution
 30 Name: Conv_gpu_0_res2_0_branch2a_1__7
 31 [0] Dest:      data0012.txt    i8[S:0.0234 O:-128][0.000,5.967]<1 x 56 x 56 x 64>
 32 [1] Src:       data0013.txt    i8[S:0.0318 O:-128][0.000,8.101]<1 x 56 x 56 x 64>
 33 [2] Filter:    data0014.txt    i8[S:0.0048 O:22][-0.718,0.502]<64 x 1 x 1 x 64>
 34 [3] Bias:      data0015.txt    i32[S:0.0002 O:0][-326470.656,326470.656]<64>
 35
 36 Type: cpumaxsplat
 37 Name: Relu_gpu_0_res2_0_branch2a_bn_2__1_max__2
 38 [0] Dest:      data0016.txt    i8[S:0.0234 O:-128][0.000,5.967]<1 x 56 x 56 x 64>
 39 [1] Src:       data0017.txt    i8[S:0.0234 O:-128][0.000,5.967]<1 x 56 x 56 x 64>
 40
 41 Type: convolution
 42 Name: Conv_gpu_0_res2_0_branch2b_1__6
 43 [0] Dest:      data0018.txt    i8[S:0.0247 O:-128][0.000,6.297]<1 x 56 x 56 x 64>
 44 [1] Src:       data0019.txt    i8[S:0.0234 O:-128][0.000,5.967]<1 x 56 x 56 x 64>
 45 [2] Filter:    data0020.txt    i8[S:0.0024 O:13][-0.332,0.269]<64 x 3 x 3 x 64>
 46 [3] Bias:      data0021.txt    i32[S:0.0001 O:0][-118353.914,118353.914]<64>

And I want to know the meaning about each line and where can I find the dataxxx.txt? Thanks a lot

**Comment by mciprian13** [comment]
@ybai62868 The file `debug.info` is printed at **compile-time**. All the dumps `dataxxx.txt` are dumped at **run-time** so you need to make an application and run the bundle in order for those dumps to be written. The files will be dumped in the same folder `debug` as the files `dataxxx.txt`.

**Comment by ybai62868** [comment]
@mciprian13 Hi, thanks for your help. I think it is a good PR and I have successfully gotten the results I want to get in the debug folder.

data0000.txt data0026.txt data0052.txt data0078.txt data0104.txt data0130.txt data0156.txt data0182.txt data0208.txt data0234.txt data0260.txt data0286.txt data0312.txt data0338.txt data0364.txt
data0001.txt data0027.txt data0053.txt data0079.txt data0105.txt data0131.txt data0157.txt data0183.txt data0209.txt data0235.txt data0261.txt data0287.txt data0313.txt data0339.txt data0365.txt
data0002.txt data0028.txt data0054.txt data0080.txt data0106.txt data0132.txt data0158.txt data0184.txt data0210.txt data0236.txt data0262.txt data0288.txt data0314.txt data0340.txt data0366.txt
data0003.txt data0029.txt data0055.txt data0081.txt data0107.txt data0133.txt data0159.txt data0185.txt data0211.txt data0237.txt data0263.txt data0289.txt data0315.txt data0341.txt data0367.txt
data0004.txt data0030.txt data0056.txt data0082.txt data0108.txt data0134.txt data0160.txt data0186.txt data0212.txt data0238.txt data0264.txt data0290.txt data0316.txt data0342.txt data0368.txt
data0005.txt data0031.txt data0057.txt data0083.txt data0109.txt data0135.txt data0161.txt data0187.txt data0213.txt data0239.txt data0265.txt data0291.txt data0317.txt data0343.txt data0369.txt
data0006.txt data0032.txt data0058.txt data0084.txt data0110.txt data0136.txt data0162.txt data0188.txt data0214.txt data0240.txt data0266.txt data0292.txt data0318.txt data0344.txt data0370.txt
data0007.txt data0033.txt data0059.txt data0085.txt data0111.txt data0137.txt data0163.txt data0189.txt data0215.txt data0241.txt data0267.txt data0293.txt data0319.txt data0345.txt data0371.txt
data0008.txt data0034.txt data0060.txt data0086.txt data0112.txt data0138.txt data0164.txt data0190.txt data0216.txt data0242.txt data0268.txt data0294.txt data0320.txt data0346.txt data0372.txt
data0009.txt data0035.txt data0061.txt data0087.txt data0113.txt data0139.txt data0165.txt data0191.txt data0217.txt data0243.txt data0269.txt data0295.txt data0321.txt data0347.txt data0373.txt
data0010.txt data0036.txt data0062.txt data0088.txt data0114.txt data0140.txt data0166.txt data0192.txt data0218.txt data0244.txt data0270.txt data0296.txt data0322.txt data0348.txt data0374.txt
data0011.txt data0037.txt data0063.txt data0089.txt data0115.txt data0141.txt data0167.txt data0193.txt data0219.txt data0245.txt data0271.txt data0297.txt data0323.txt data0349.txt data0375.txt
data0012.txt data0038.txt data0064.txt data0090.txt data0116.txt data0142.txt data0168.txt data0194.txt data0220.txt data0246.txt data0272.txt data0298.txt data0324.txt data0350.txt debug.info
data0013.txt data0039.txt data0065.txt data0091.txt data0117.txt data0143.txt data0169.txt data0195.txt data0221.txt data0247.txt data0273.txt data0299.txt data0325.txt data0351.txt
data0014.txt data0040.txt data0066.txt data0092.txt data0118.txt data0144.txt data0170.txt data0196.txt data0222.txt data0248.txt data0274.txt data0300.txt data0326.txt data0352.txt
data0015.txt data0041.txt data0067.txt data0093.txt data0119.txt data0145.txt data0171.txt data0197.txt data0223.txt data0249.txt data0275.txt data0301.txt data0327.txt data0353.txt
data0016.txt data0042.txt data0068.txt data0094.txt data0120.txt data0146.txt data0172.txt data0198.txt data0224.txt data0250.txt data0276.txt data0302.txt data0328.txt data0354.txt
data0017.txt data0043.txt data0069.txt data0095.txt data0121.txt data0147.txt data0173.txt data0199.txt data0225.txt data0251.txt data0277.txt data0303.txt data0329.txt data0355.txt
data0018.txt data0044.txt data0070.txt data0096.txt data0122.txt data0148.txt data0174.txt data0200.txt data0226.txt data0252.txt data0278.txt data0304.txt data0330.txt data0356.txt
data0019.txt data0045.txt data0071.txt data0097.txt data0123.txt data0149.txt data0175.txt data0201.txt data0227.txt data0253.txt data0279.txt data0305.txt data0331.txt data0357.txt
data0020.txt data0046.txt data0072.txt data0098.txt data0124.txt data0150.txt data0176.txt data0202.txt data0228.txt data0254.txt data0280.txt data0306.txt data0332.txt data0358.txt
data0021.txt data0047.txt data0073.txt data0099.txt data0125.txt data0151.txt data0177.txt data0203.txt data0229.txt data0255.txt data0281.txt data0307.txt data0333.txt data0359.txt
data0022.txt data0048.txt data0074.txt data0100.txt data0126.txt data0152.txt data0178.txt data0204.txt data0230.txt data0256.txt data0282.txt data0308.txt data0334.txt data0360.txt
data0023.txt data0049.txt data0075.txt data0101.txt data0127.txt data0153.txt data0179.txt data0205.txt data0231.txt data0257.txt data0283.txt data0309.txt data0335.txt data0361.txt
data0024.txt data0050.txt data0076.txt data0102.txt data0128.txt data0154.txt data0180.txt data0206.txt data0232.txt data0258.txt data0284.txt data0310.txt data0336.txt data0362.txt
data0025.txt data0051.txt data0077.txt data0103.txt data0129.txt data0155.txt data0181.txt data0207.txt data0233.txt data0259.txt data0285.txt data0311.txt data0337.txt data0363.txt

The only question is that what the meaning of each part?

**Comment by mciprian13** [comment]
The file `debug.info` tells you which file corresponds to which IR instruction operand. Glow has multiple levels as you might know:
- Graph level (high level) with `nodes`
- IR level (low level) with `instructions`
The dump correspond to the second low-level IR. Between the high-level IR and low-level IR in most cases there is a 1:1 mapping but there are some `node` which are broken into multiple `instructions`. 

When you compile the model using `model-compiler` you can get more information about each level:
- you can use the flag `-dump-graph-DAG=graph.dot` to dump the graph-level representation in a DOT file. Next you can use the Linux utility `dot` to convert the DOT file into a PDF file like this: `dot -Tpdf graph.dot -o graph.pdf`
- you can use the flag `-g` to dump the IR text representation in a file `model.glow` in the same folder as the bundle artifacts

Based on this two pieces of information you can make the association between each buffer/file and the node/instruction operand it corresponds to.
